File: python/parallelMonte.py
import sys
import multiprocessing as mp
from multiprocessing import Pool

def worker(args):
    thread_id, per_thread = args
    seed = thread_id + 12345
    hits = 0
    a = 1664525
    c = 1013904223
    m = 2147483647
    for _ in range(per_thread):
        seed = (a * seed + c) & 0x7fffffff
        x = seed / m
        seed = (a * seed + c) & 0x7fffffff
        y = seed / m
        if x * x + y * y <= 1.0:
            hits += 1
    return hits

if __name__ == "__main__":
    # Только на Linux!
    mp.set_start_method('fork')

    threads = int(sys.argv[1])
    total = int(sys.argv[2])
    per_thread = total // threads

    with Pool(processes=threads) as pool:
        results = pool.map(worker, [(t, per_thread) for t in range(threads)])

    total_hits = sum(results)
    pi = 4.0 * total_hits / (per_thread * threads)
    print(pi)

File: perl/parallelMonte.pl
use strict;
use warnings;
use threads;

my ($threads, $total) = @ARGV;
my $per_thread = int($total / $threads);

my @thr;
for my $t (0..$threads-1) {
    push @thr, threads->create(sub {
        my $seed = $t + 12345;
        my $hits = 0;
        for (1..$per_thread) {
            $seed = ($seed * 1664525 + 1013904223) & 0x7fffffff;
            my $x = $seed / 2147483647.0;
            $seed = ($seed * 1664525 + 1013904223) & 0x7fffffff;
            my $y = $seed / 2147483647.0;
            $hits++ if $x*$x + $y*$y <= 1.0;
        }
        return $hits;
    });
}

my $total_hits = 0;
$total_hits += $_->join() for @thr;

my $pi = 4.0 * $total_hits / ($per_thread * $threads);
print "$pi\n";